mdsuite/calculators/angular_distribution_function.py

```diff
@@ -158,105 +218,186 @@ def _load_positions(self, indices: list) -> tf.Tensor:
         loaded_data : tf.Tensor
                 tf.Tensor of tensor_values loaded from the hdf5 database_path
         """
-
-        if self.molecules:
-            path_list = [join_path(species, "Positions") for species in self.experiment.molecules]
-            return self.experiment.load_matrix("Positions", path=path_list, select_slice=np.s_[:, indices])
+        path_list = [join_path(species, "Positions") for species in self.species]
+        data = self.experiment.load_matrix("Positions", path=path_list, select_slice=np.s_[:, indices])
+        if len(self.species) == 1:
+            return data
         else:
-            return self.experiment.load_matrix("Positions", select_slice=np.s_[:, indices])
+            return tf.concat(data, axis=0)
 
-    def run_experimental_analysis(self):
+    def _prepare_data_structure(self):
         """
-        Perform the ADF analysis
+        Prepare variables and dicts for the analysis.
+        Returns
+        -------
+
         """
         sample_configs = np.linspace(self.start, self.stop, self.n_confs, dtype=np.int)
 
         species_indices = []
         start_index = 0
         stop_index = 0
-        for species in self.experiment.species:
+        for species in self.species:
             stop_index += len(self.experiment.species.get(species).get('indices'))
-            species_indices.append(
-                (species, start_index, stop_index)
-            )
+            species_indices.append((species, start_index, stop_index))
             start_index = stop_index
 
-        angles = {}
+        return sample_configs, species_indices
 
-        # TODO select when to enable tqdm of get_triplets
+    def _prepare_generators(self, sample_configs):
+        """
+        Prepare the generators and compiled functons for the analysis.
 
+        Returns
+        -------
+
+        """
+
+        def generator():
+            for idx in sample_configs:
+                yield self._load_positions(idx)
+
+        return generator
+
+    def _prepare_triples_generator(self):
+        """
+        Prepare the triples generators including tf.function
+        Returns
+        -------
+
+        """
         if self.use_tf_function:
-            @tf.function
+            @tf.function()
```

Comment:
I think that the brackets aren't required

Reply:
Oh yes sorry, I was experimenting with the compile and the graph_mode stuff.